contrib/rust/src/python/pants/contrib/rust/tasks/cargo_toolchain.py

```diff
@@ -0,0 +1,143 @@
+    std_out = workunit.output('stdout')
+    std_err = workunit.output('stderr')
+
+    try:
+      subprocess.check_call(command, shell=True, stdout=std_out, stderr=std_err)
+    except subprocess.CalledProcessError as e:
+      workunit.set_outcome(1)
+      std_err.write('Execution failed: {0}'.format(e))
+    workunit.set_outcome(3)
+
+  def download_rustup_install_script(self):
+    return requests.get('https://sh.rustup.rs')
+
+  def check_integrity_of_rustup_install_script(self, install_script):
+    hasher = hashlib.sha256(install_script.encode('utf-8'))
+    current_fingerprint = hasher.hexdigest() if PY3 else hasher.hexdigest().decode('utf-8')
+    if current_fingerprint != self.LAST_KNOWN_FINGERPRINT:
+      raise TaskError(
+        'The fingerprint of the rustup script has changed!\nLast known: {0}\ncurrent: {1}'.format(
+          self.LAST_KNOWN_FINGERPRINT, current_fingerprint))
+
+  def install_rust_toolchain(self, toolchain):
+    self.context.log.info('Installing toolchain: {0}'.format(toolchain))
+    with self.context.new_workunit(name='install-rustup-toolchain',
+                                   labels=[WorkUnitLabel.BOOTSTRAP]) as workunit:
+      cmd = ['rustup', 'install', toolchain]
+
+      env = {
+        'PATH': (self.context.products.get_data('cargo_env')['PATH'], True)
+      }
+
+      self.run_command(cmd, get_buildroot(), env, workunit)
+
+      if workunit.outcome() != WorkUnit.SUCCESS:
+        self.context.log.error(workunit.outcome_string(workunit.outcome()))
+      else:
+        self.context.log.info(workunit.outcome_string(workunit.outcome()))
+
+  def set_new_toolchain_as_default(self, toolchain):
+    self.context.log.info('Set {0} as default toolchain'.format(toolchain))
+    with self.context.new_workunit(name='setup-rustup-toolchain',
+                                   labels=[WorkUnitLabel.BOOTSTRAP]) as workunit:
+      cmd = ['rustup', 'default', toolchain]
+
+      env = {
+        'PATH': (self.context.products.get_data('cargo_env')['PATH'], True)
+      }
+
+      self.run_command(cmd, get_buildroot(), env, workunit)
+
+      if workunit.outcome() != WorkUnit.SUCCESS:
+        self.context.log.error(workunit.outcome_string(workunit.outcome()))
+      else:
+        self.context.log.info(workunit.outcome_string(workunit.outcome()))
+
+  @staticmethod
+  def check_if_rustup_exist():
+    return distutils.spawn.find_executable('rustup') is not None
+
+  def setup_toolchain(self):
+    toolchain = self.get_options().toolchain
+    if toolchain:
+      self.install_rust_toolchain(toolchain)
+      self.set_new_toolchain_as_default(toolchain)
+
+  def set_cargo_path(self):
+    env = os.environ.copy()
+    self.context.products.safe_create_data('cargo_env', lambda: {})
+    cargo_env = self.context.products.get_data('cargo_env')
+    cargo_env['PATH'] = os.path.join(env['HOME'], '.cargo/bin')
+
+  def execute(self):
+    self.context.log.info('Check if rust toolchain exist.')
```

Comment:
Anything that will run for every re-compile after an edit should probably be at debug level: these two messages probably qualify.